rafthttp/peer.go

```diff
@@ -15,158 +15,196 @@
 package rafthttp
 
 import (
-	"errors"
+	"fmt"
+	"log"
 	"net/http"
-	"sync"
 	"time"
 
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
 	"github.com/coreos/etcd/etcdserver/stats"
 	"github.com/coreos/etcd/pkg/types"
 	"github.com/coreos/etcd/raft/raftpb"
 )
 
 const (
-	appRespBatchMs = 50
-	propBatchMs    = 10
-
+	DialTimeout      = time.Second
 	ConnReadTimeout  = 5 * time.Second
 	ConnWriteTimeout = 5 * time.Second
-)
 
-type peer struct {
-	sync.Mutex
+	recvBufSize = 4096
 
-	id  types.ID
-	cid types.ID
+	streamTypeGeneral = "General"
+	streamTypeMsgApp  = "MsgApp"
+)
 
-	tr http.RoundTripper
-	// the url this sender post to
-	u  string
+type peer struct {
+	to types.ID
 	r  Raft
-	fs *stats.FollowerStats
 
-	batcher     *Batcher
-	propBatcher *ProposalBatcher
-
-	pipeline *pipeline
-	stream   *stream
-
-	paused  bool
-	stopped bool
+	msgAppDialer  *streamDialer
+	msgAppReader  *streamReader
+	msgAppDecoder *msgAppDecoder
+	msgAppWriter  *streamWriter
+	dialer        *streamDialer
+	reader        *streamReader
+	writer        *streamWriter
+	pipeline      *pipeline
+
+	sendc   chan raftpb.Message
+	recvc   chan raftpb.Message
+	updatec chan string
+	pausec  chan struct{}
+	resumec chan struct{}
+	stopc   chan struct{}
+	done    chan struct{}
 }
 
-func NewPeer(tr http.RoundTripper, u string, id types.ID, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
-	return &peer{
-		id:          id,
-		cid:         cid,
-		tr:          tr,
-		u:           u,
-		r:           r,
-		fs:          fs,
-		pipeline:    newPipeline(tr, u, id, cid, fs, errorc),
-		stream:      &stream{},
-		batcher:     NewBatcher(100, appRespBatchMs*time.Millisecond),
-		propBatcher: NewProposalBatcher(100, propBatchMs*time.Millisecond),
+func startPeer(tr http.RoundTripper, u string, local, to, cid types.ID, r Raft, fs *stats.FollowerStats, errorc chan error) *peer {
+	p := &peer{
+		to:            to,
+		r:             r,
+		msgAppDialer:  newStreamDialer(u, tr, local, to, cid, streamTypeMsgApp),
+		msgAppDecoder: newMsgAppDecoder(local, to),
+		msgAppWriter:  startStreamWriter(newMsgAppEncoder(to, fs)),
+		dialer:        newStreamDialer(u, tr, local, to, cid, streamTypeGeneral),
+		writer:        startStreamWriter(&generalEncoder{}),
+		pipeline:      newPipeline(tr, u, to, cid, fs, errorc),
+
+		sendc:   make(chan raftpb.Message),
+		recvc:   make(chan raftpb.Message, recvBufSize),
+		updatec: make(chan string),
+		pausec:  make(chan struct{}),
+		resumec: make(chan struct{}),
+		stopc:   make(chan struct{}),
+		done:    make(chan struct{}),
 	}
+	p.msgAppReader = startStreamReader(p.msgAppDialer, p.msgAppDecoder, p.recv)
```

Comment:
It reads from the remote by itself. :wink: 
Yep, it behaves much like server-client mode now, considering they use channel to communicate now.
I would stay for a while(of course, before merging it) to see whether it is the best naming.